### ISSUE TITLE
Allow Transformations to be aware of a "fixed frame"

### ISSUE THREAD

**Issue by jleibs** [rerun-io/rerun#1522]
Rerun currently only looks up transform-chains using the latest-at value for the given scene.

However, there are many situations where this doesn't produce the correct results. In particular, the case sensor-readings that are different from an odometry update the judder artifacts are quite clear:
[Screencast from 02-27-2023 07_58_23 PM.webm](https://user-images.githubusercontent.com/3312232/223319830-0dc49582-87b3-435d-847b-1aae3e15fbf5.webm)

The crux of the problem is the points are observed in the robot frame. So to place them in the map frame, we apply the transform `map_T_robot`. However, `map_T_robot` is being updated at a much faster rate (based on odometry) and so our current "latest at" logic ends up using points from time t with multiple transforms observed at t+1, t+2, etc.

The right thing to do here is to instead build up the transform chain specifically using historical transform values at the time of the data being transformed (including interpolation if the robot odometry measurements and point-cloud measurements didn't line up exactly).

However, it's important to note that this isn't *always* the right behavior. Whether transforms are treated as having a fixed frame needs to be configurable. This seems like a good candidate for an EntityProperty.  If the "fixed frame" entity property is set, rather than using the current latest-at behavior, we would decompose the transform chain into two pieces:
```
dest_T_source = dest_T_fixed @ T_render * fixed_T_source @ T_source
```

The solution for this should also take into account visible history: https://github.com/rerun-io/rerun/issues/723

**Comment by alcolado** [comment]
Is there any plan or timeline for this? I see it mentioned in #1537. This "Judder" seems to be a very common artifact of all kinds of visualizers, it would be cool if Rerun had a built-in way to fix it. I'm also interested in knowing any suggestions for work-arounds.

I have come across the same kind of problem, and I have a snippet here to illustrate it (the link to the video in the original post no longer works for me, but I have watched it in the past). 

### My Reproduction

```python
"""
A mouse chases a cat.

The cat has a sensor which tells us where it is in the world at a high frequency.

The mouse detector tells us where the mouse is in the world, but this detector runs at lower frequency.

updates over time:
cat_pos     x x x x x x x x x x x x x x x x
mouse_pos   x         x         x         x

If we look at the mouse in the cat frame, it will appear that the cat catches up to the mouse before the
mouse suddenly jumps away again. This is because the latest cat position updates bring us closer to the mouse
before a new mouse detection occurs.

Instead, we might want to use the cat position at the time the mouse position was logged. This produces a smoother
view of the mouse in the cat's frame.
"""

import rerun as rr
import rerun.blueprint as rrb

rr.init("cat_mouse")

world_view = rrb.Spatial3DView(
    name="World",
    origin="/ground",
    contents=["/ground", "/cat", "/mouse"],
    eye_controls=rrb.EyeControls3D(
        kind="Orbital",
        look_target=(0.0, 0.5, 0.0),
        position=(0.0, -20.0, 20.0)
    )
)
cat_view = rrb.Spatial3DView(
    name="POV Cat",
    origin="/cat", contents=["/ground", "/cat", "/mouse"],
    eye_controls=rrb.EyeControls3D(
        kind="Orbital",
        look_target=(0.0, 0.0, 0.0),
        position=(0.0, -12.0, 12.0)
    )
)
cat_view2 = rrb.Spatial3DView(
    name="POV Cat (Relative Mouse)",
    origin="/cat", contents=["/ground", "/cat", "/mouse_relative_to_cat"],
    eye_controls=rrb.EyeControls3D(
        kind="Orbital",
        look_target=(0.0, 0.0, 0.0),
        position=(0.0, -12.0, 12.0)
    )
)

blueprint = rrb.Horizontal(world_view, cat_view, cat_view2)

rr.spawn(default_blueprint=blueprint)

v_cat = 0.1
v_mouse = 0.11

ground = rr.LineStrips3D([[[-5, i, 0], [5, i, 0]] for i in range(20)])
cat = rr.Boxes3D(half_sizes=(1, 1, 1), centers=(0, 0, 1), colors=(255, 0, 0), fill_mode="solid")
mouse = rr.Points3D(positions=(0, 0, 1), radii=(1,), colors=(0, 0, 255))

rr.log("/ground", rr.CoordinateFrame("world_frame"), ground, static=True)
rr.log("/cat", rr.CoordinateFrame("cat_frame"), cat, static=True)

for t in range(100):
    rr.set_time("time", sequence=t)

    # world positions
    pos_cat = v_cat * t
    pos_mouse = v_mouse * t + 5

    # cat sensor tells me where the cat is at a high frequency
    rr.log("/tf", rr.Transform3D(translation=(0, pos_cat, 0), parent_frame="world_frame", child_frame="cat_frame"))

    # mouse detector runs less frequently
    if t % 10 == 0:
        rr.log("/mouse", rr.CoordinateFrame("world_frame"), rr.InstancePoses3D(translations=(0, pos_mouse, 0)), mouse)

        # In general, this would require resolving a transformation tree
        pos_mouse_relative_to_cat = pos_mouse - pos_cat
        rr.log("/mouse_relative_to_cat",
               rr.CoordinateFrame("mouse_relative_to_cat_frame"),
               rr.Transform3D(
                   translation=(0, pos_mouse_relative_to_cat, 0),
                   parent_frame="cat_frame",
                   child_frame="mouse_relative_to_cat_frame"),
               mouse)



rr.set_time()
```
This produces the what I show in the video below. There is a cat chasing a mouse. The cat position is updated at a fast rate, but the mouse detections happen more slowly. In my case, the mouse detections are in the world frame and the "Judder" shows up if I transform to the cat frame (middle panel). What I would hope to see instead (if I want) is on the right panel. That is, for viewing the mouse in the cat frame, I could chose to only use the tf at the time the mouse was detected. It's okay to use the latest tf for the other entities (like the background floor). 

I think this is completely analogous to the original post (although in that case, it was viewing it in the world that caused the judder, and it was many lidar points instead of a mouse)

https://github.com/user-attachments/assets/f2c47425-e922-43b8-9144-6733a977ba73

### Possible solutions I considered 
1. Interpolating the mouse data to match the cat data (this could be expensive)
2. Downsample the cat data to match the mouse rate (I lose smoothness in the motion between cat and background)
3. Transform the cat into the mouse frame before logging (repeats work that Rerun already could be doing)
4. Create a new frame with the transformation frozen in time (based on a comment in #11965). This is what I did in the snippet. This is simple if I already have the transform, but maybe in general it requires resolving a transformation tree (something which rerun already knows how to do). 

So seeing how Rerun already knows the transformations over time, already knows how to resolve the transformation tree, and already knows how to transform my entities, I wish I could just tap into rerun to do this 😆. 

### A solution that required forking Rerun
For my own use, I played around with making new archetype called `rr.FrozenTransform`. It works like this
```python
rr.log("/mouse", rr.CoordinateFrame("mouse_detection"), mouse) # mouse is some Points3D, or boxes, or whatever...
rr.log("/tf", rr.FrozenTransform(parent_frame="cat", child_frame="world", frozen_frame="mouse_detection")  # mouse detection frame only has a connection to cat when it's logged, internally this gets resolved to the cat-world transform
```
Internally, I needed to modify `TransformForrest::new` to detect this `FrozenTransform` and resolve it (by building a new `TransformTree` between parent and child at the logging time instead of "latest" time). All the changes were in `re_tf` and the visualizers all seem to work properly since in the end it just behaves like another coordinate frame. In case it's of interest, these are the changes https://github.com/rerun-io/rerun/compare/main...alcolado:rerun:alcolado/frozen-tf

### Solutions that don't require fork
I think if I created an .rrd file, I could make a DerivedLense that does something similar, but maybe I would still have to work out the transform tree myself? I haven't tried this yet. Also, in my use case I would like for it to work kind of "live", rather than making a second pass on a recorded file.  

### Dream solution
As a more general solution, It would be cool if there was just a button you could click in the UI to toggle on or off using latest tf/tf at logged time (and per entity I guess). 

### Summary
Anyway, in summary 
* I'm expressing interest in a feature that solves this
* Looking for any suggested solutions
* Looking for feedback on my solution(s)
* Happy to contribute in any way, whether it be examples/use cases/feedback/code

**Comment by MichaelGrupp** [comment]
This behavior can be also illustrated in a more realistic scenario with a simple modification of the [ARKit Scenes example](https://github.com/rerun-io/rerun/tree/main/examples/python/arkit_scenes) - by downsampling the camera log calls w.r.t. the transform updates.

The depth camera's point cloud appears drifting in between log timestamps of the camera, due to the camera being attached to the moving frame and the point cloud's transform being re-resolved w.r.t the fixed target frame at every time cursor tick.

https://github.com/user-attachments/assets/e900ce32-2735-4fae-8e3d-1378aaa782e2

**Comment by MichaelGrupp** [comment]
A maybe simpler option that doesn't require to specify fixed frames for each entity or re-build a transform tree at every time tick:

* use the view's target frame as reference (blueprint setting, like now)
* introduce two transform resolution modes in the viewer (should be also a blueprint setting)
    * latest-at global time cursor (**current behavior)**: re-compute the transform at every time cursor tick, using the latest-at state of the transform tree at that tick
    * latest-at data time (**new**): compute the transform of an entity only at the time it gets updated using the latest-at transform tree at that log timestamp, cache that transform until data updates again

That new mode would essentially "freeze" 3D data w.r.t. the target frame in the viewer after it first appears, until a new data timestamp occurs.

For example RViz, as a simpler live visualizer, renders data like pointclouds like that. E.g. when a laser scan arrives, that scan is placed w.r.t. RViz's fixed frame (target frame) using a transform lookup at the data timestamp, and not re-transformed using the wall clock later.

---

Some things to consider:
* Rerun's visible time range settings might need special handling to work with this, or have some restrictions.
* What component triggers an update?
    * Rerun's component model allows partial updates of entities, and maybe not every partial update should update trigger a transform update (e.g. color component of a point cloud).
    * This could be potentially extra fields for `CoordinateFrame`
        * pro: fits because `CoordinateFrame` describes semantics of the transform
        * contra: this might be awkward for users of implicit entity-path transforms

(there's probably more to take into account, just noting down ideas)

**Comment by alcolado** [comment]
Yes, the two transform resolution modes as a blueprint settings would be perfect!

**Comment by alcolado** [comment]
"What component triggers an update?" is a good point. I wonder if components can be classified by "spatial" or "not spatial"? Maybe this is too ambiguous, or would require too much documentation and special handling...

Alternatively, I think it could be acceptable that any partial update at all would trigger a transform update. 
* pro: If the partial update (conceptually speaking) was due to "this was the only thing that changed", then the trigger makes sense (we could have equivalently done a full update). 
* con: If the partial update was due to "this was the only thing that was measured" (and we didn't measure anything about the position), then I guess we are back to the original problem of creating judder. 

Just thinking out loud...

**Comment by Daniiiil1** [comment]
The two-mode blueprint proposal maps cleanly to a fixed-frame transform split. For a sample logged at t_data and a viewer cursor at t_view:

target_T_source = target_T_fixed(t_view) * fixed_T_source(t_data)

This freezes the measured geometry in the fixed frame at acquisition time while still allowing the view/target side of the chain to move at render time.

The important implementation detail is cache identity. With visible history, this cannot be cached only per entity: different visible rows of the same entity may require different transforms. I would key the resolved value by at least timeline, payload row id/data time, source frame, target/fixed frame, and transform-graph generation. The existing per-timeline transform cache and row ids appear to provide most of those primitives.

For interpolation, translation can be linear and rotation SLERP with quaternion hemisphere correction. I would avoid extrapolation across missing intervals or transform resets by default.

For partial updates, “any component update” can recreate the judder when only color or metadata changes. A more stable rule is to anchor transform time to the row containing the primary spatial payload/pose, or introduce an explicit spatial timestamp/anchor rather than infer it from every component update.

Would the team prefer that explicit anchor semantic, or an archetype-defined set of spatial components? That decision seems worth fixing before the cache/API work.

**Comment by Wumpf** [comment]
For starters I'd aggressively cache the entire `re_tf::TransformForest` in a simplistic cache since you need to walk the tree anyways to get a transform for a given tree. Yes that's inefficient and may break down with lots of entities at different update intervals and when we add time ranges. But for all so far described scenarios that would be perfectly viable.
What triggers the update should be configurable with an enum such that we can extend it in the future. The easiest to implement by far will be to update on a given local component identifier because then one can just look up the last time at which that component has been logged and use that time to lookup (or compute just-in-time) the correct `re_tf::TransformForest`.

Also no interpolation for the first iteration. Looks easy at first but it would need look-ahead and handling various error cases.

**Comment by MichaelGrupp** [comment]
I think it's also worth to note that in our case, it's probably more logical that the view target frame ***is*** the fixed frame. This extra distinction between fixed and target frame makes sense for a generic transform lookup API like `tf2_ros` (see the "advanced" TF API [here](https://github.com/ros2/geometry2/blob/f6053126926a38ffad5e81588054d793d87fc662/tf2_ros/include/tf2_ros/buffer.hpp#L149)), which has to support individual transform lookup requests at arbitrary times.

Here we are primarily interested in the viewer use case, where all objects are rendered w.r.t. the view target frame. With a new "latest-at data time" mode, we would resolve each rendered entity's transform to the target frame at the data time ([see also previous comment](https://github.com/rerun-io/rerun/issues/1522#issuecomment-5090966924)).

So that extra fixed frame notion doesn't make sense to me in addition to the existing view target frame, for the viewer case. Not sure if I overlooked something ;)

**Comment by Daniiiil1** [comment]
I implemented the narrowed first iteration from the discussion in a fresh draft PR: https://github.com/rerun-io/rerun/pull/12878

The draft uses the view target frame as the reference frame, adds `GlobalTimeCursor` and `ComponentTime` modes, anchors component-time resolution to a configured local component identifier, caches complete `TransformForest` snapshots per latest-at query, and does not interpolate.

I also kept the missing-component behavior explicit: that entity falls back to global-cursor resolution. The new test covers the original judder scenario, a time-varying coordinate frame, and that fallback.

The two points where feedback would be most useful are whether the trigger component should remain view-wide for the first version, and whether the fallback semantics are appropriate. I left per-visible-row transforms for time-range rendering outside this MVP.

**Comment by alcolado** [comment]
I've been thinking about this a little bit more, and believe that using the target frame as the fixed frame may not be sufficient (I initially thought it would be). Perhaps the best path forward would be to support the full lookup from tf2 https://docs.ros2.org/latest/api/tf2_ros/classtf2__ros_1_1Buffer.html#a98127244155ee66c933b080270ecb05d
```
lookupTransform(target_frame, target_time, source_frame, source_time, fixed_frame)
```

<img width="1114" height="404" alt="Image" src="https://github.com/user-attachments/assets/ec7d0fea-b411-445e-9efc-e5dffca6d6c6" />

The `target_time` is the "current time" on the timeline. Currently, rerun is supporting `fixed_frame=source_frame`, so the `source_time` doesn't matter. Using "latest at data time" mode would be equivalent to `fixed_frame=target_frame` (so the `source_time` is all that matters).

<img width="1114" height="404" alt="Image" src="https://github.com/user-attachments/assets/05752234-55c4-4242-89a6-5bfafd08b7cb" />

But I can image scenarios where you want to view the data from a frame which is not the natural "fixed frame" of the data. Additionally, if you allow different views with their own different target frames to all be fixed frames of the data, there may no longer be a consistent position of the data across all views (your "hover over" location in one view could potentially not match another view). 

Instead, I think the "fixed frame" should be attached to the data/entity path in a similar way to the coordinate frame. It could be an optional argument to  `rr.CoordinateFrame`, e.g. `rr.CoordinateFrame("lidar", fixed_frame="world")`, or it could be a separate archetype. By default it would fall back to the coordinate frame of the entity path, which replicates the current behavior today.

## Example
This is a "toy example", but it is similar to a real scenario. There is a lidar on a moving robot. If the lidar updates are slower than the robot position updates, there will be "judder" (the points are cutting into the detected object and then jumping back after the new lidar update). 

```python
# at a high frequency
rr.log("/tf", rr.Transform3D(parent_frame="world", child_frame="robot", ...))
...
# at a lower frequency
rr.log("/lidar", rr.Points3D(points_3d), rr.CoordinateFrame("robot"))
```

https://github.com/user-attachments/assets/b36e34b6-27e9-4e3c-981a-ca8e69e16774

So to fix this you would want to say the the "fixed frame" of the lidar points is the "world".
```python
rr.log("/lidar",  rr.Points3D(points_3d), rr.CoordinateFrame("robot", fixed_frame="world"))  # hypothetical new fixed_frame argument
```
 
https://github.com/user-attachments/assets/5fd4d35e-1525-45f7-a0b0-45955ae83793

In the example I'm showing, I made a (contrived) observer frame, just to illustrate that the fixed frame is not necessarily the view target frame. If it were the fixed frame, you would introduce a different judder like so:

https://github.com/user-attachments/assets/c218276c-c50e-4d99-a7c0-259138096107

## Caching
My understanding is that there is a "TransformationForest" computed once per frame an re-used by all the visualizers. With "fixed frames", this would still be shared for all the `target time` transformation lookups. However, there would be many different `source_time` trees needed for any data that uses fixed frames. Perhaps these could be computed "on the fly" and cached for lookup within the same frame? I'm not sure if this is what was being suggested in a previous comment.

## Visible History
Transformation Trees at different times are also needed for "visible history". With visible history, the `source_time` is when the archetype appeared, and the `target_time` should be the "last time this archetype was visible". I think the "last time is was visible" makes sense to make the history continuous in time (for arcehtypes that are still visible, the last time it was visible is "now"). 

I don't know if this was even suggested, but I don't know whether new transformation archetypes should trigger copies of other archetypes (e.g. the same object is repeated for every transform update to make a potentially very dense trail). Just stating this as a possible interpretation of the visible history... 

## Code
This is the code to generate my toy example if anyone cares
```python
import numpy as np
import rerun as rr
import rerun.blueprint as rrb

view = rrb.Spatial3DView(
    contents=["/**"],
    spatial_information=rrb.SpatialInformation(target_frame="observer"),
    eye_controls=rrb.EyeControls3D(
        position=(15, -25, 30),
        look_target=(10, 0, 0)
    )
)

rr.init("lidar-jitter")
rr.spawn(default_blueprint=view)

N = 50_000
speed = 0.05


for i in range(250):
    rr.set_time("frame", sequence=i)

    lidar_x = i * speed

    # make a lidar cone
    lidar_depth = 20
    points = np.random.rand(N, 2)
    points[:, 0] *= lidar_depth
    points[:, 1] = (points[:, 1] * 2 - 1) * lidar_depth / 2
    x, y = points.T
    invisible = np.abs(y) / x > 0.5

    points = points[~invisible]
    points[:, 0] += lidar_x

    # pretend there is an object at x = 15 with half size 2
    x0 = 15
    s = 2
    x, y = points.T
    invisible = s / (x0 - lidar_x) > np.abs(y) / (x - lidar_x)
    invisible = invisible & (x > x0)
    points = points[~invisible]
    rr.log("/box", rr.Boxes3D(centers=(x0+s, 0, 0), half_sizes=(s, s, s)), rr.CoordinateFrame("world"), static=True)


    rr.log("/robot", rr.Points3D((0, 0, 0), radii=1), rr.CoordinateFrame("robot"))
    rr.log("/tf", rr.Transform3D(translation=(lidar_x, 0, 0), parent_frame="world", child_frame="robot"))

    # some contrived other frame
    rr.log("/observer", rr.Transform3D(translation=(0, i*speed, 0), parent_frame="world", child_frame="observer"))

    if i % 25 == 0:
        # go back into lidar frame
        points[:, 0] -= lidar_x
        points_3d = np.concatenate([points, np.zeros_like(points[:, [0]])], axis=1)
        rr.log("/lidar", rr.Points3D(points_3d))
        rr.log("/lidar", rr.CoordinateFrame("robot"))
```

I computed the "fixed frame" alternatives with my own "lookupTransform" code that isn't worth sharing...

## Conclusion
I think implementing the full tf2 style lookup would make it possible to handle visible history and fixed frames.
